package/standalone_file_logger/standalone_file_logger/src/rotating_logger.cc

```diff
@@ -187,35 +187,65 @@ double RotatingLogger::get_time_passed()
 
 void RotatingLogger::frame_handler(const uint8_t *data, size_t size)
 {
-  if (!_dest_available) {
-    // check imediately on startup for path availability. Subsequently, check
-    // periodically
-    if (_session_start_time.time_since_epoch().count() != 0 && get_time_passed() < _poll_period) {
-      return;
+  // TODO: keep track of total data bytes: enforce a limit/warning in .h
+  std::vector<uint8_t> *frame = new std::vector<uint8_t> (data, data + size);
+
+  std::unique_lock<std::mutex> mlock(_mutex);
+  _queue.push_back( frame );
+  mlock.unlock();
+  _cond.notify_one();
+}
+
+void RotatingLogger::process_frame()
+{
+  std::unique_lock<std::mutex> mlock(_mutex, std::defer_lock);
+  while (!_finished) {
+    if (!_dest_available) {
+      // check imediately on startup for path availability. Subsequently, check
+      // periodically
+      if (_session_start_time.time_since_epoch().count() != 0 && get_time_passed() < _poll_period) {
+        return;
+      }
+      _session_start_time = std::chrono::steady_clock::now();
+      if (!start_new_session()) {
+        return;
+      }
     }
-    _session_start_time = std::chrono::steady_clock::now();
-    if (!start_new_session()) {
+    if (!check_slice_time()) {
       return;
     }
-  }
-  if (!check_slice_time()) {
-    return;
-  }
 
-  size_t num_written = 0;
-  if (_cur_file != nullptr) {
-    num_written = fwrite(data, 1, size, _cur_file);
-  }
-  if (num_written != size) {
-    // If drive is removed needs to close file imediately and not attempt to
-    // open new file for a couple seconds to avoid locking mount
-    close_current_file();
-    _dest_available = false;
-    // wait _poll_period to check drive again
-    _session_start_time = std::chrono::steady_clock::now();
-    log_msg(LOG_WARNING, std::string("Write to file failed: ") + strerror(errno));
+    mlock.lock();
+    if (_queue.empty()) {
+      _cond.wait(mlock);
+      // recheck !_finished
+      mlock.unlock();
+      continue;
+    }
+
+    size_t num_written = 0;
+    std::vector<uint8_t> *frame = _queue.front();
+    size_t size = sizeof(std::vector<uint8_t>::value_type) * frame->size();
+    if (_cur_file != nullptr) {
+      num_written = fwrite(&frame[0], sizeof(std::vector<uint8_t>::value_type), frame->size(), _cur_file);
+    }
+
+    _queue.pop_front();
```

Comment:
I think this code should move above the call to `fwrite`:
```c
     _queue.pop_front(); 
     mlock.unlock();
```

Since this would block additions to the queue while writing.

Reply:
fixed